File: weles/cloudflare/challenge.py
"""Cloudflare challenge detection and bypass using vision.

Uses Claude vision to detect the challenge and locate the verification
checkbox. The click is dispatched through CDPMouse which uses Bezier
curves and realistic timing by default.

If Claude refuses to answer the vision query (safety refusal), the
underlying VisionRefusedError propagates so the failure is visible
and the prompt or context can be fixed.
"""

import logging

from .config import CF_CHECK_INTERVAL_MS

logger = logging.getLogger(__name__)


async def wait_cloudflare(page, timeout_ms: int = 72000,
                          settle_ms: int = 5000) -> bool:
    """Wait for Cloudflare challenge to resolve via vision-based clicking.

    Raises VisionRefusedError if Claude refuses to identify the click
    target. Returns True if no challenge or challenge clears within
    timeout, False if the challenge does not clear before timeout.
    """
    from ..cdp.dom.vision import check_page, find_click_target, ask_page

    await page.wait_for_timeout(settle_ms)

    raw_answer = await ask_page(
        page,
        "Is this a Cloudflare security verification or challenge page? "
        "Answer only YES or NO.",
    )
    is_cf = raw_answer.strip().upper().startswith("YES")
    print(f"  [cloudflare] raw vision answer: {repr(raw_answer)}")
    print(f"  [cloudflare] challenge detected: {is_cf}")

    if not is_cf:
        return True

    target = await find_click_target(
        page, "the checkbox or button to verify you are human")
    print(f"  [cloudflare] click target: {target}")
    if target is None:
        print("  [cloudflare] no click target found - challenge likely "
              "in non-interactive auto-pass mode")
    else:
        await page.mouse.click(target["x"], target["y"])
        print(f"  [cloudflare] clicked at ({target['x']}, {target['y']})")

    checks = timeout_ms // CF_CHECK_INTERVAL_MS
    for i in range(checks):
        await page.wait_for_timeout(CF_CHECK_INTERVAL_MS)
        still_cf = await check_page(
            page,
            "Is this a Cloudflare security verification or challenge page?",
        )
        print(f"  [cloudflare] check {i+1}/{checks}: still challenged = "
              f"{still_cf}")
        if not still_cf:
            return True

    return False


async def is_challenged(page) -> bool:
    """Check if the page is currently showing a Cloudflare challenge."""
    from ..cdp.dom.vision import check_page
    return await check_page(
        page,
        "Is this a Cloudflare security verification or challenge page?",
    )


async def bypass_cloudflare(page, solver=None, timeout_ms: int = 72000) -> bool:
    """Full Cloudflare bypass via vision-based detection and clicking."""
    return await wait_cloudflare(page, timeout_ms=timeout_ms)

File: src/cloudflare/challenge.ts
/**
 * Cloudflare challenge detection and bypass — 1:1 port of weles/cloudflare/challenge.py
 *
 * Uses Claude vision to detect the challenge and locate the verification
 * checkbox. The click is dispatched through CDPMouse with Bezier curves.
 */

import { askPage, checkPage, findClickTarget, type ScreenshottablePage } from '../vision/analyze.js';

const CF_CHECK_INTERVAL_MS = 3000;

// Fast-path DOM check: real Cloudflare challenge pages always contain one of
// these strings in title/body. If none match, skip the vision call entirely
// (askClaude shells out to the LLM router and adds 5-15s per page on every
// goto, even on platforms that never serve CF). Verified 2026-05-02:
// LinkedIn /feed/ on stale cookies stalled WSession.goto for 70+ s in this
// path because LinkedIn doesn't use Cloudflare and the vision call hung.
async function looksLikeCloudflareDom(page: any): Promise<boolean> {
  try {
    const r = await page.evaluate(() => {
      const t = (document.title || '').toLowerCase();
      const b = (document.body?.innerText || '').slice(0, 600).toLowerCase();
      const hasCfMarker = /cloudflare|just a moment|attention required|checking your browser|verifying you are human|enable javascript and cookies/.test(t + ' ' + b);
      const hasCfFrame = !!document.querySelector('iframe[src*="challenges.cloudflare.com"], iframe[src*="cdn-cgi/challenge-platform"]');
      return hasCfMarker || hasCfFrame;
    });
    return !!r;
  } catch { return false; }
}

export async function waitCloudflare(page: any, timeoutMs = 72000, settleMs = 5000): Promise<boolean> {
  await page.waitForTimeout(settleMs);

  // Cheap DOM probe before the expensive vision call. If the page has zero
  // Cloudflare-shaped markers, return true immediately (treated as "not
  // challenged"). Saves a round-trip per goto on every non-CF platform.
  if (!(await looksLikeCloudflareDom(page))) return true;

  const rawAnswer = await askPage(
    page as ScreenshottablePage,
    'Is this a Cloudflare security verification or challenge page? Answer only YES or NO.',
  );
  const isCf = rawAnswer.trim().toUpperCase().startsWith('YES');
  console.log(`  [cloudflare] raw vision answer: ${JSON.stringify(rawAnswer)}`);
  console.log(`  [cloudflare] challenge detected: ${isCf}`);

  if (!isCf) return true;

  const target = await findClickTarget(
    page as ScreenshottablePage,
    'the checkbox or button to verify you are human',
  );
  console.log(`  [cloudflare] click target: ${JSON.stringify(target)}`);
  if (!target) {
    console.log('  [cloudflare] no click target found - challenge in auto-pass mode');
  } else {
    await page.mouse.click(target.x, target.y);
    console.log(`  [cloudflare] clicked at (${target.x}, ${target.y})`);
  }

  const checks = Math.floor(timeoutMs / CF_CHECK_INTERVAL_MS);
  for (let i = 0; i < checks; i++) {
    await page.waitForTimeout(CF_CHECK_INTERVAL_MS);
    const stillCf = await checkPage(
      page as ScreenshottablePage,
      'Is this a Cloudflare security verification or challenge page?',
    );
    console.log(`  [cloudflare] check ${i + 1}/${checks}: still challenged = ${stillCf}`);
    if (!stillCf) return true;
  }

  return false;
}

export async function isChallenged(page: any): Promise<boolean> {
  return checkPage(
    page as ScreenshottablePage,
    'Is this a Cloudflare security verification or challenge page?',
  );
}

export async function bypassCloudflare(page: any, timeoutMs = 72000): Promise<boolean> {
  return waitCloudflare(page, timeoutMs);
}

export type { ScreenshottablePage as CFPage };
